state/store.go

```diff
@@ -228,132 +229,158 @@ func (store dbStore) Bootstrap(state State) error {
+	// height but only when they change or at a check point, it is likely that the validator set at the height
+	// we prune to is empty and thus dependent on the validator set saved at a previous height. We must find
+	// that validator set and make sure it is not pruned.
+	lastRecordedValSetHeight := lastStoredHeightFor(height, valInfo.LastHeightChanged)
+	lastRecordedValSet, err := loadValidatorsInfo(store.db, lastRecordedValSetHeight)
+	if err != nil || lastRecordedValSet.ValidatorSet == nil {
+		return fmt.Errorf("couldn't find validators at height %d (height %d was originally requested): %w",
+			lastStoredHeightFor(height, valInfo.LastHeightChanged),
+			height,
+			err,
+		)
+	}
+
+	// batch delete all the validators sets up to height
+	return store.batchDelete(
+		validatorsKey(1),
+		validatorsKey(height),
+		validatorsKey(lastRecordedValSetHeight),
+	)
+}
+
+// pruneConsensusParams calls a reverse iterator from base height to retain height batch deleting
+// all consensus params in between. If the consensus params at the new base height is dependent
+// on a prior height then this will keep that lower height to.
+func (store dbStore) pruneConsensusParams(retainHeight int64) error {
+	paramsInfo, err := store.loadConsensusParamsInfo(retainHeight)
+	if err != nil {
+		return fmt.Errorf("consensus params at height %v not found: %w", retainHeight, err)
 	}
-	keepParams := make(map[int64]bool)
+
+	// As we don't save the consensus params at every height, only when there is a consensus params change,
+	// we must not prune (or save) the last consensus params that the consensus params info at height
+	// is dependent on.
 	if paramsInfo.ConsensusParams.Equal(&tmproto.ConsensusParams{}) {
-		keepParams[paramsInfo.LastHeightChanged] = true
+		lastRecordedConsensusParams, err := store.loadConsensusParamsInfo(paramsInfo.LastHeightChanged)
+		if err != nil || lastRecordedConsensusParams.ConsensusParams.Equal(&tmproto.ConsensusParams{}) {
+			return fmt.Errorf(
+				"couldn't find consensus params at height %d as last changed from height %d: %w",
+				paramsInfo.LastHeightChanged,
+				retainHeight,
+				err,
+			)
+		}
 	}
 
+	// batch delete all the consensus params up to the retain height
+	return store.batchDelete(
+		consensusParamsKey(1),
+		consensusParamsKey(retainHeight),
+		consensusParamsKey(paramsInfo.LastHeightChanged),
+	)
+}
+
+// pruneABCIResponses calls a reverse iterator from base height to retain height batch deleting
+// all abci responses in between
+func (store dbStore) pruneABCIResponses(height int64) error {
+	return store.batchDelete(abciResponsesKey(1), abciResponsesKey(height), nil)
+}
+
+// batchDelete is a generic function for deleting a range of keys in reverse order. It will
+// skip keys that have been
+func (store dbStore) batchDelete(start []byte, end []byte, exception []byte) error {
+	iter, err := store.db.ReverseIterator(start, end)
```

Comment:
why reverse iterator here?

Reply:
Because e.g. validator sets have dependencies on previous validator sets (via `LastHeightChanged`), so if we delete the previous sets first then loading later sets may error/panic.